backend/cfg/llvmize.ml

```diff
@@ -0,0 +1,574 @@
+
+let fun_attrs _t _codegen_options : string list =
+  (* CR gyorsh: translate and communicate to llvm backend *)
+  []
+
+(* Mark regs we need to allocate *)
+let collect_regs t (cfg : Cfg.t) =
+  let body_regs =
+    Cfg.fold_blocks cfg
+      ~f:(fun _ block regs ->
+        let body_regs =
+          DLL.fold_left block.body
+            ~f:(fun regs (instr : Cfg.basic Cfg.instruction) ->
+              Reg.Set.add_seq (Array.to_seq instr.res) regs)
+            ~init:Reg.Set.empty
+        in
+        let terminator_regs = Array.to_seq block.terminator.res in
+        Reg.Set.add_seq terminator_regs body_regs |> Reg.Set.union regs)
+      ~init:Reg.Set.empty
+  in
+  let arg_regs = Array.to_seq cfg.fun_args |> Reg.Set.of_seq in
+  let temp_regs = Reg.Set.diff body_regs arg_regs in
+  !(t.current_fun_info).arg_regs <- arg_regs;
+  !(t.current_fun_info).temp_regs <- temp_regs
+
+let alloca_regs t =
+  let arg_regs = !(t.current_fun_info).arg_regs in
+  let temp_regs = !(t.current_fun_info).temp_regs in
+  (* At this point, [t] will have mapped the args to idents. We will be using
+     the alloca'd idents instead, so we save them and make [t] forget about the
+     old ones. *)
+  let old_arg_idents =
+    Reg.Set.to_seq arg_regs |> List.of_seq
+    |> List.map (fun reg -> reg, get_ident_for_reg t reg)
+  in
+  Reg.Tbl.clear !(t.current_fun_info).reg2ident;
+  Reg.Set.iter (F.alloca t) arg_regs;
+  Reg.Set.iter (F.alloca t) temp_regs;
+  List.iter (fun (reg, old_ident) -> F.ins_store t old_ident reg) old_arg_idents
+
+let cfg (cl : CL.t) =
+  let t = Option.get !current_compilation_unit in
+  reset_fun_info t;
+  let layout = CL.layout cl in
+  let cfg = CL.cfg cl in
+  (* CR gyorsh: handle unboxed return type (where is this info? do we need to
+     find [Return] instruction to find it? if so, add a field to [Cfg.t] to
+     store it explicitly. *)
+  let { Cfg.blocks;
+        fun_name;
+        fun_args;
+        fun_codegen_options;
+        fun_dbg;
+        entry_label;
+        fun_contains_calls = _ (* not used at this point *);
+        fun_num_stack_slots = _ (* only available after regalloc *);
+        fun_poll = _ (* not needed after poll insertion *);
+        next_instruction_id = _
+      } =
+    cfg
+  in
+  let pp_block label =
+    let block = Label.Tbl.find blocks label in
+    let preds = Cfg.predecessor_labels block in
+    if Label.equal entry_label label && not (Misc.Stdlib.List.is_empty preds)
+    then Misc.fatal_errorf "Entry label must not have predecessors";
+    F.block_label_with_predecessors t label preds;
+    DLL.iter ~f:(F.basic t) block.body;
+    F.terminator t block.terminator
+  in
+  let pp_body () =
+    (* last argument id + 1 is reserved, so we cannot use it. we explicitly skip
+       it here *)
+    let (_ : Ident.t) = Ident.Gen.get_fresh !(t.current_fun_info).ident_gen in
+    alloca_regs t;
+    F.ins t "br %a" (F.pp_label t) entry_label;
```

Comment:
I think this branch is not needed (because entry_label is always first in the layout), but no harm.